### PR TITLE
Document autoscaling (resolves #1225)

### LINKED ISSUE #1225: Add documentation for new cluster related utils

### PR DESCRIPTION
**Pull request by cket**

### CODE REVIEW

**Comment by cket** [comment]
@hannes-ucsc ready

File: docs/installation.rst [hunk 1]
```diff
@@ -231,8 +231,9 @@ Cloud installation
 
 Installation on AWS for distributed computing
 ---------------------------------------------
-We use CGCloud_ to provision instances and clusters in AWS. Thorough documentation of CGCloud_ can be found
-in the CGCloud-core_ and CGCloud-toil_ documentation. Brief steps will be provided to those
+We use CGCloud_ and the Toil AWS provisioner to provision instances and clusters in AWS. Thorough documentation
```
Comment (hannes-ucsc) [inline docs/installation.rst]: Give more importance to the AWS provisioner. CGCloud is on the way out, mention it last.

File: docs/running.rst [hunk 2]
```diff
@@ -317,6 +317,46 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
 
 On all cloud providers, it is recommended that you run long-running jobs on remote systems under ``screen``. Simply type ``screen`` to open a new ``screen` session. Later, type ``ctrl-a`` and then ``d`` to disconnect from it, and run ``screen -r`` to reconnect to it. Commands running under ``screen`` will continue running even when you are disconnected, allowing you to unplug your laptop and take it home without ending your Toil jobs.
 
+.. _Autoscaling:
+
+Autoscaling
+-----------
+The fastest way to get started runnning Toil in a cloud environment is using Toil's autoscaling capabilities to
+handle node provisioning for us. Currently, autoscaling is only supported on the AWS cloud platform with two choices
+of provisioners: CGCloud and Toil's own Docker based provisioner.
```
Comment (hannes-ucsc) [inline docs/running.rst]: Same here.

File: docs/running.rst [hunk 3]
```diff
@@ -317,6 +317,46 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+
+The AWS provisioner is included in Toil alongside the [aws] extra and allows us to spin up a cluster without any external dependencies
+using the Toil Appliance docker image. To get started, we choose which version of the appliance to run from CGL's
+`Quay account <https://quay.io/repository/ucsc_cgl/toil-leader?tab=tags>`_ and set the TOIL_APPLIANCE_SELF environment
+variable to quay.io/ucsc-cgl:toil-- followed by the tag of the Toil Appliance version we wish to use::
+
+    export TOIL_APPLIANCE_SELF=quay.io/ucsc_cgl/toil:3.5.0a1--80c340c5204bde016440e78e84350e3c13bd1801
```
Comment (hannes-ucsc) [inline docs/running.rst]: Can you not hardcode the commit and instead give users a way to find the "latest" image. If not, this is fine, we just need to remember to take this out once #1246 lands.

**Comment by cket** [comment]
@hannes-ucsc ready for review

File: docs/running.rst [hunk 4]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
 On all cloud providers, it is recommended that you run long-running jobs on remote systems under ``screen``. Simply type ``screen`` to open a new ``screen` session. Later, type ``ctrl-a`` and then ``d`` to disconnect from it, and run ``screen -r`` to reconnect to it. Commands running under ``screen`` will continue running even when you are disconnected, allowing you to unplug your laptop and take it home without ending your Toil jobs.
 
+.. _Autoscaling:
+
+Autoscaling
+-----------
+The fastest way to get started runnning Toil in a cloud environment is using
+Toil's autoscaling capabilities to handle node provisioning for us.
+Currently, autoscaling is only supported on the AWS cloud platform with two
+choices of provisioners: Toil's own Docker based provisioner and CGCloud.
```
Comment (hannes-ucsc) [inline docs/running.rst]: Docker-based

File: docs/running.rst [hunk 5]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+
+The AWS provisioner is included in Toil alongside the [aws] extra and allows
+us to spin up a cluster without any external dependencies using the Toil
+Appliance docker image. To get started, we choose which version of the
```
Comment (hannes-ucsc) [inline docs/running.rst]:
```
using the Toil Appliance, a Docker image that bundles Toil and all its requirements, e.g. Mesos.
```

Or something like that. Once `Toil Appliance` is introduced, I would switch to using `appliance` for short.

File: docs/running.rst [hunk 6]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+appliance to run from CGL's
+`Quay account <https://quay.io/repository/ucsc_cgl/toil-leader?tab=tags>`_and
+ set the TOIL_APPLIANCE_SELF environment variable to quay.io/ucsc-cgl:toil--
```
Comment (hannes-ucsc) [inline docs/running.rst]: backquote environment variable names.

Comment (hannes-ucsc) [inline docs/running.rst]: and values

File: docs/running.rst [hunk 7]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+ followed by the tag of the Toil Appliance version we wish to use ::
```
Comment (hannes-ucsc) [inline docs/running.rst]: no space before `::`

File: docs/running.rst [hunk 8]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+    --keyPairName=your-AWS-key-pair-name
+    _
+
+to launch a t2.micro leader instance - you should adjust this instance type
+accordingly if you want to do real work. See 
+`here <https://aws.amazon.com/ec2/instance-types/>`_ for a full selection
+ of EC2 instance types. For more information on cluster managment using 
+ Toil's AWS provisioner, see :ref:`clusterRef`.
+
+To use CGCloud based autoscaling, see :ref:`installationAWS`for CGCloud
```
Comment (hannes-ucsc) [inline docs/running.rst]: CGCloud-based

File: docs/running.rst [hunk 9]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+installation and more information on starting our leader instance.
+
+Once we have our leader instance launched, the steps for both provisioners
```
Comment (hannes-ucsc) [inline docs/running.rst]: I think elsewhere in this document we tend to use 2nd person, e.g. "you and your cluster". I don't know if we should be consistent. Might not be a big deal.

File: docs/running.rst [hunk 10]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+converge. At this point we simply start off our Toil run, being sure to
+set ``--batchSystem=mesos``, using an AWS job store with the following
```
Comment (hannes-ucsc) [inline docs/running.rst]: s/set/pass

File: docs/running.rst [hunk 11]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+autoscaling specific options. Indicate your provisioner choice via the
```
Comment (hannes-ucsc) [inline docs/running.rst]: The sentence that ends on this line reads awkwardly.

File: docs/running.rst [hunk 12]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+``--provisioner=<>`` flag and node type for your worker nodes via
+``--nodeType=<>``. Addionally, both provisioners support
+`preemptable nodes <https://aws.amazon.com/ec2/spot/>`_. Toil can run on a
+heterogenous cluster of both preemptable and non preemptable nodes. Our
+preemptable node type can be set by using the ``--preemptableNodeType=<>``
+flag. Be sure to read the help strings for this option, as the different
```
Comment (hannes-ucsc) [inline docs/running.rst]: Mention `--help`

File: docs/running.rst [hunk 13]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+provisioners have different syntax for indicating the bid price for on the
+node type you want. While individual jobs can each explicitly specify whether
```
Comment (hannes-ucsc) [inline docs/running.rst]: They do?

File: docs/running.rst [hunk 14]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+or not they should be run on preemptable nodes, the ``--defaultPreemptable``
```
Comment (hannes-ucsc) [inline docs/running.rst]: mention `preemptable=` keyword. Mention that its a requirement like disk and cores etc.

File: docs/running.rst [hunk 15]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+flag will allow all other jobs to run on preemptable machines. Finally,
```
Comment (hannes-ucsc) [inline docs/running.rst]: … all jobs that don't specify the `preemptable=` requirement …

File: docs/running.rst [hunk 16]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+we can set the maximum number of preemptable and non-preemptable nodes
+via the flags ``--maxNodes=<>`` and ``--maxPreemptableNodes=<>``.
```
Comment (hannes-ucsc) [inline docs/running.rst]: Mention the possibility of creating a hang here. Mention overflow, or deficit compensation. I forgot what the command line option is called but sth like --preemptableDeficitCompensation.

File: docs/running.rst [hunk 17]
```diff
@@ -317,6 +317,57 @@ There are several recommended ways to run Toil jobs in the cloud. Of these, runn
...
+Appliance docker image. To get started, we choose which version of the
+appliance to run from CGL's
+`Quay account <https://quay.io/repository/ucsc_cgl/toil-leader?tab=tags>`_and
+ set the TOIL_APPLIANCE_SELF environment variable to quay.io/ucsc-cgl:toil--
+ followed by the tag of the Toil Appliance version we wish to use ::
+
+    export TOIL_APPLIANCE_SELF=quay.io/ucsc_cgl/toil:3.5.0a1--80c340c5204bde016440e78e84350e3c13bd1801
+    _
+
+Having done that, simply run::
```
Comment (hannes-ucsc) [inline docs/running.rst]: How would the user determine the exact value for TOIL_APPLIANCE_SELF?

File: docs/running.rst (hunk 13 above)
Comment (cket) [inline docs/running.rst]: Yeah, I saw it in the help string here: https://github.com/BD2KGenomics/toil/blob/master/src/toil/common.py#L348
That's why the AWS provisioner uses c3.8xlarge:0.42 vs CGCloud's c3.8xlarge{|:0.42}

**Comment by cket** [comment]
@hannes-ucsc ready for review

**Pull request merged**